daemon/src/rollover/maker.rs

```diff
@@ -0,0 +1,293 @@
+
+use crate::command;
+use crate::oracle;
+use crate::rollover::protocol::*;
+use crate::setup_contract;
+
+use super::protocol;
+
+type ListenerConnection = (
+    Framed<Substream, JsonCodec<ListenerMessage, DialerMessage>>,
+    PeerId,
+);
+
+/// Permanent actor to handle incoming substreams for the `/itchysats/rollover/1.0.0`
+/// protocol.
+///
+/// There is only one instance of this actor for all connections, meaning we must always spawn a
+/// task whenever we interact with a substream to not block the execution of other connections.
+pub struct Actor {
+    tasks: Tasks,
+    oracle_pk: schnorrsig::PublicKey,
+    get_announcement: Box<dyn MessageChannel<oracle::GetAnnouncement>>,
+    n_payouts: usize,
+    pending_protocols: HashMap<OrderId, ListenerConnection>,
+    executor: command::Executor,
+}
+
+impl Actor {
+    pub fn new(
+        executor: command::Executor,
+        oracle_pk: schnorrsig::PublicKey,
+        get_announcement: Box<dyn MessageChannel<oracle::GetAnnouncement>>,
+        n_payouts: usize,
+    ) -> Self {
+        Self {
+            tasks: Tasks::default(),
+            oracle_pk,
+            get_announcement,
+            n_payouts,
+            pending_protocols: HashMap::default(),
+            executor,
+        }
+    }
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = ();
+
+    async fn stopped(self) -> Self::Stop {}
+}
+
+#[xtra_productivity(message_impl = false)]
+impl Actor {
+    async fn handle(&mut self, msg: NewInboundSubstream, ctx: &mut xtra::Context<Self>) {
+        let NewInboundSubstream { peer, stream } = msg; // TODO: Use `PeerId` for something!
+        let address = ctx.address().expect("we are alive");
+
+        self.tasks.add_fallible(
+            async move {
+                let mut framed =
+                    Framed::new(stream, JsonCodec::<ListenerMessage, DialerMessage>::new());
+
+                let propose = framed
+                    .next()
+                    .await
+                    .context("End of stream while receiving Propose")?
+                    .context("Failed to decode Propose")?
+                    .into_propose()?;
+
+                address
+                    .send(ProposeReceived {
+                        propose,
+                        framed,
+                        peer,
+                    })
+                    .await?;
+
+                anyhow::Ok(())
+            },
+            |e| async move { tracing::warn!("Failed to handle incoming rollover protocol: {e:#}") },
+        );
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle(&mut self, msg: ProposeReceived) {
+        let ProposeReceived {
+            propose,
+            framed,
+            peer,
+        } = msg;
+        let order_id = propose.order_id;
+
+        if let Err(e) = self
+            .executor
+            .execute(order_id, |cfd| {
+                // TODO: Validate that the correct peer is invoking this?
```

Comment:
Yeah, this is just about failing early. A taker could "attack" a maker by sending bogus rollover proposals to keep the maker busy. No money would be lost, but it could result in a denial of service attack. Not super important yet but if we can easily do it it might be better to just do it now :)